### ISSUE TITLE
[consistent-type-imports] ignores `TSTypeReference` reports false positive for "used by decorator metadata"

### ISSUE THREAD

**Issue by tquetano-netflix** [typescript-eslint/typescript-eslint#8582]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

Apologies in advance for the complex bug report, but I could not reproduce in the playground, and attempts at getting a repro outside of our massive repository have proven fruitless. Hopefully the detailed explanation can suffice.

When applying the `consistent-type-imports` rule, I received a number of false positives that something is being used as decorator metadata when it was only being used as a type import. Spelunking in the rule code, I discovered it was falling back to the ultimate case [here](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/consistent-type-imports.ts#L237).

I traced it back to where the type was passed as a generic value. Simple examples:

```ts
const [state, setState] = useState<TypeOnlyImport>(false);
const wm = new WeakMap<TypeOnlyImport, boolean>();
```

In these cases, the ancestor tree looked like this:
- `TSTypeReference`
- `TSTypeParameterInstantiation`
- `CallExpression`

None of these are handled in [the `switch` statement](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/consistent-type-imports.ts#L237), hence falling back to it being seen as a value. As a test, I added another `case` handler:

```js
case utils_1.AST_NODE_TYPES.TSTypeReference:
  return true;
```

And all passes! I'm happy to submit a PR, but wanted to surface this for feedback in case there was something else in the configuration that might be the cause.

### Reproduction Repository Link

N/A

### Repro Steps

N/A

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `7.1.0` |
| `@typescript-eslint/parser`            | `7.1.0` |
| `@typescript-eslint/scope-manager`     | `7.1.0` |
| `@typescript-eslint/typescript-estree` | `7.1.0` |
| `@typescript-eslint/type-utils`        | `7.1.0` |
| `@typescript-eslint/utils`             | `7.1.0` |
| `TypeScript`                           | `5.3.3` |
| `ESLint`                               | `8.57.0` |
| `node`                                 | `20.11.1` |

**Comment by bradzacher** [comment]
> I received a number of false positives that something is being used as decorator metadata

This is a known bug in the rule - it falls back to this error in cases where it doesn't know what's going on. But the cases that occurs are rare and are generally indicative of the problems.

----

Given this does not reproduce in the [playground](https://typescript-eslint.io/play/#ts=4.9.5&fileType=.tsx&code=JYWwDg9gTgLgBAbwFB1XArgZwKYGUYCGM2ANCmgCoCeY2A8gHYA2VAkuNDGQPTdwACMTAFpgAcwbRsSAL5wAZlAgg4AIhAQAJqoDcSJAGMIDTPADapoqTg4Y%2BKwF04AXgw57xADzVajFu0hYAD4ACnkCJhwASj0jE3gAdxVXBmwEuAB1bAIAawBZAjBvGnpmNg5YEjgAIwgIJmyGUJikIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) I'm inclined to believe that this isn't a bug in the rule and is a bug somewhere else.

The rule is a single-file lint rule which means that it doesn't use types and doesn't consume any data across files. So if it doesn't repro in the playground either there's config missing to make it repro, or there's some form of version mismatch.

Based on my gut I'd lean toward the latter - there's likely a mismatch in parser version and plugin version. You can usually check this by searching your lock file(s) for references to `@typescript-eslint/`. If there are multiple versions there then that can cause these sorts of issues. Each major we change our AST and some releases we fix bugs in rules and the scope analyser. So if you have multiple majors and/or minors installed you fall into a world of "undefined" behaviour.

Could you please try creating an isolated reproduction repo for this? Personally I go about this by spinning up a GH repo and copying across the package.json and eslint config file, then add an example file with a failing case from the codebase. Then I delete lines from the config until it no longer reproduces. Then I similarly delete lines from the package.json.

**Comment by tquetano-netflix** [comment]
@bradzacher It appears you are correct ... due to nested dependencies, it does appear that there are multiple versions at play. I'm digging in to see how I can mitigate this because this is upstream. Appreciate the context, this helped.
